### ISSUE TITLE
Online cropping of images?

### ISSUE THREAD

**Issue by jramapuram** [NVIDIA/DALI#88]
Is it possible to use DALI to return something like a lambda, eg:

```python
# for each image in a batch return something like this
img = ... 
lbda = lambda x, y, width, height: img.crop(x, y, x+width, y+height)
```

I have a use case where my model (pytorch) spits out `x`, `y`, `width` and `height` and would like to use DALI to produce an image crop. In my current setup with `PIL-SIMD` this is causing a pretty large bottleneck as `PIL-SIMD` takes a while to crop the image, transfer the data to GPU memory and then continue model execution. 

Workflow:

    img0 --> [x, y, w, h] --> crop in host memory --> transfer to GPU memory --> img1 --> model execution --> model output

Would DALI be able to provide some speedups for me here? Or is this a hopeless endeavor?

**Comment by ptrendx** [comment]
It is not really possible with current DALI, but we are working on improvements for `ExternalSource` op which could enable such use case. Basically, I don't think we currently can accept pointer on the GPU as external input for the pipeline, but there should not be any technical reason why we could not add such functionality.
@Kh4L any thoughts?

Tracked internally as DALI-182

**Comment by cliffwoolley** [comment]
With Unified Memory, the source could conceptually be anywhere accessible to the processor that handles the next op in the pipeline, right?  Sysmem, peer mem...

**Comment by jramapuram** [comment]
Not sure the mmap type solution is applicable (if I understand that correctly). My end goal is working with 100k x 100k images. Ideally (if possible) it would be useful to load just header information and crop in real-time with minimal latency. To be honest, I’m not sure this is even plausible.

I’m currently testing a rust based implementation based on piston/image and rayon via python ffi calls, but was hoping for something a little more performant with a gpu impl.

**Comment by ptrendx** [comment]
Hmmm, I'm not sure then how your system is structured - in the first post you said that you have a model that takes an image and produces `[x,y,width,height]` - how do you want to fit such a large image in GPU mem then in the first place? Or is the `[x,y,width,height]` generated from some smaller version of the image?

**Comment by jramapuram** [comment]
Good question; your assumption is correct. The coordinates are generated by a related image which is much smaller :)

**Comment by ptrendx** [comment]
Is that a single large image that needs to be cropped? Or are there multiple such images? If there are multiple images, then basically what you request is a decode with ROI (region of interest), so that you do not have to decode the whole image potentially (although I'm not sure if this is even viable with formats like JPEG).

If this is a single image and you have it decoded already, then I'm not sure why cropping would be a bottleneck - what is the size of the crop?

**Comment by jramapuram** [comment]
It's a mini-batch of images. True, not all formats do support it, but JPEG does, so does BMP for example. I do not think PNG does though.

As stated, the problem setting is that the ROI is extracted using something else that the model parses: this is initially variably sized, but then resized through a `bilinear` transform to a fixed size (say [32x32]). The entire original image isn't decoded; it's header information containing the dimensions / # of channels, etc are merely what is extracted.

Been trying to do something like this using Rust here: https://github.com/jramapuram/parallel_image_crop but would be awesome if something like this were plausible using DALI

**Comment by rceballos98** [comment]
Has there been any updates on this? I am very interested in being able to decode only a small region of in a large image at a time using DALI. 

I saw tiff was recently added to the list of supported image types. 
Tiff makes this easier than JPEG since its tiled and memory mapped. 

Essentially, I would need functionality like openslide's read_region that given a TIFF or SVS image will only decode the ROI. 

Like so:
`read_region(location, level, size)`
Return an RGBA Image containing the contents of the specified region.

Parameters: 
- location (tuple) : (x, y) tuple giving the top left pixel in the level 0 reference frame
- level (int) : the level number
- size (tuple) – (width, height) tuple giving the region size

(from https://openslide.org/api/python/)

Do you think I could accomplish this with an external operation? 
Are there plans to support this in the future?

**Comment by JanuszL** [comment]
Hi,
This is somewhere on our long term roadmap but we cannot commit to any date. We would be more than happy to accept any external contribution implementing it. Recently we have added that for JPEG images - https://github.com/NVIDIA/DALI/pull/503, other images are fully decoded and then necessary window is extracted. Supporting this for other images is a matter of the decoder providing such fucntionality.
Regarding TIFF support - we can decode those images but as it was said, no true ROI support is implemented.

**Comment by JanuszL** [comment]
Hi,
In the current master and soon in 0.16 you can use the ExternalSource operator to provide additional input data to other operators, like Slice or DecoderSlice. I think it should do want you to ask for.

**Comment by JanuszL** [comment]
This change is available in [DALI 0.16.0](https://github.com/NVIDIA/DALI/releases/tag/v0.16.0).
